src/libsyntax/ext/base.rs

```diff
@@ -146,6 +156,173 @@ impl Into<Vec<Annotatable>> for Annotatable {
+                   -> TokenStream;
+}
+
+impl<F> AttrProcMacro for F
+    where F: Fn(TokenStream, TokenStream) -> TokenStream
+{
+    fn expand<'cx>(&self,
+                   _ecx: &'cx mut ExtCtxt,
+                   _span: Span,
+                   annotation: TokenStream,
+                   annotated: TokenStream)
+                   -> TokenStream {
+        // FIXME setup implicit context in TLS before calling self.
+        (*self)(annotation, annotated)
+    }
+}
+
+pub struct TokResult<'a> {
+    pub parser: Parser<'a>,
+    pub span: Span,
+}
+
+impl<'a> MacResult for TokResult<'a> {
+    fn make_items(mut self: Box<Self>) -> Option<SmallVector<P<ast::Item>>> {
+        if self.parser.sess.span_diagnostic.has_errors() {
+            return None;
+        }
+
+        let mut items = SmallVector::zero();
+        loop {
+            match self.parser.parse_item() {
+                Ok(Some(item)) => {
+                    // FIXME better span info.
+                    let mut changer = ChangeSpan { span: self.span };
+                    let new_items = changer.fold_item(item);
+                    items.push_all(new_items);
+                }
+                Ok(None) => {
+                    return Some(items);
+                }
+                Err(mut e) => {
+                    e.emit();
+                    return Some(SmallVector::zero());
+                }
+            }
+        }
+    }
+
+    fn make_impl_items(mut self: Box<Self>) -> Option<SmallVector<ast::ImplItem>> {
+        let mut items = SmallVector::zero();
+        loop {
+            match self.parser.parse_impl_item() {
+                Ok(item) => {
+                    // FIXME better span info.
+                    let mut changer = ChangeSpan { span: self.span };
+                    let new_items = changer.fold_impl_item(item);
+                    items.push_all(new_items);
+
+                    return Some(items);
+                }
+                Err(mut e) => {
+                    e.emit();
+                    return Some(SmallVector::zero());
+                }
+            }
+        }
+    }
+
+    fn make_trait_items(mut self: Box<Self>) -> Option<SmallVector<ast::TraitItem>> {
+        let mut items = SmallVector::zero();
+        loop {
+            match self.parser.parse_trait_item() {
+                Ok(item) => {
+                    // FIXME better span info.
+                    let mut changer = ChangeSpan { span: self.span };
+                    let new_items = changer.fold_trait_item(item);
+                    items.push_all(new_items);
+
+                    return Some(items);
+                }
+                Err(mut e) => {
+                    e.emit();
+                    return Some(SmallVector::zero());
+                }
+            }
+        }
+    }
+
+    fn make_expr(mut self: Box<Self>) -> Option<P<ast::Expr>> {
+        match self.parser.parse_expr() {
+            Ok(e) => Some(e),
```

Comment:
Something like `ensure_complete_parse` is also needed here so that `valid[expression] } followed by arbitrary tokens` is not allowed (c.f. my comment about `ParserAnyMacro`).